### ISSUE TITLE
Biquad-family frequency responses take nfft-point FFTs of 3-tap polynomials (and a latent NaN in |H| backward)

### ISSUE THREAD

**Issue by SebastianJiroSchlecht** [gdalsanto/flamo#35]
## Summary

Every biquad-family filter computes its frequency response by taking an `nfft`-point `rfft` of a **3-tap** polynomial. At `nfft = 2**17` that pads three coefficients with 131069 zeros, per channel, per numerator and denominator. FFTs are ~25 % of self CPU time in an FDN training step.

## Where

The same two lines appear in 12 classes in `flamo/processor/dsp.py` (24 call sites): `Biquad`, `parallelBiquad`, `SOSFilter`, `parallelSOSFilter`, `SVF`, `parallelSVF`, `GEQ`, `parallelGEQ`, `PEQ`, `parallelPEQ`, `AccurateGEQ`, `parallelAccurateGEQ`.

```python
b_aa = torch.einsum("p, pon -> pon", self.alias_envelope_dcy, b)
a_aa = torch.einsum("p, pon -> pon", self.alias_envelope_dcy, a)
B = torch.fft.rfft(b_aa, self.nfft, dim=0)
A = torch.fft.rfft(a_aa, self.nfft, dim=0)
```

Since `rfft(x, n=nfft)[m] = Σ_p x[p]·exp(-2πi·m·p/nfft)`, a 3-tap transform is a contraction against a `(3, nfft//2+1)` grid — `P` multiply-adds per bin rather than an `O(nfft log nfft)` transform per channel. The grid depends only on `nfft`, so it caches.

## Measurements

Response call alone, `parallelSOSFilter`, `N = 16`, `nfft = 2**17`, float32, grad enabled: **57.0 ms → 12.4 ms (4.6×)**.

In a full FDN training step (`N = 16`, `nfft = 2**17`, CPU float32): 291.6 → 241.2 ms/step, **1.21×**.

But the win is very shape-dependent. Across all 12 classes at `nfft = 2**15` with 2–4 channels the median is only **1.2×**, and `Biquad` in float64 with 8 channels is **0.4× — 2.5× slower**, because the FFT is already cheap there while the grid contraction still costs `P·M`. Any real patch wants a guard on tap count and probably on channel count too.

## The reason I have not sent a patch

Replacing the FFT changes the rounding, and in one case that surfaces **NaN gradients where the current code has none**:

```
Biquad, lowpass, alias_decay_db=60, float32, nfft=2**15

stock    B: exact zeros 7/131080   H: exact zeros 4/65540   grad NaNs 0/16
grid     B: exact zeros 0/131080   H: min|H| 3.19e-42       grad NaNs 4/16
```

At a spectral null the FFT path's rounding lands on **exactly** zero, and `abs()`'s backward at exact zero returns the subgradient 0. The grid path resolves the true value instead — a float32 **denormal**, below the 1.18e-38 normal minimum — and `abs()`'s backward then divides by it and produces NaN.

So the direct evaluation is the *more* accurate of the two, and that is precisely what breaks it. Checked against a float64 reference of the same path:

| | stock f32 vs f64 | grid f32 vs f64 | f64: stock vs grid |
|---|---|---|---|
| `GEQ` | 4.1e-03 | **1.8e-03** | 1.0e-11 |
| `parallelSOSFilter` | 1.1e-05 | **9.0e-06** | 6.0e-14 |
| `PEQ` | 1.8e-06 | **1.2e-06** | 2.4e-15 |

(The float32 numbers look alarming for `GEQ` but that is float32's own error band on a cascade of ~10 sections — the two paths simply sit on opposite sides of it.)

This suggests a latent issue independent of any optimisation: `get_poly_coeff`'s guard replaces exactly-zero *denominators* with `eps`, but nothing protects `|H|` from being denormal-but-nonzero, and a loss on `|H|` will then NaN. Worth handling on its own terms, and it is a prerequisite for changing the transform.

Happy to send the grid evaluation as a PR — with a tap/channel guard and the near-zero clamp — if you agree with that direction.

## Environment

flamo 0.2.18 (`de29cac`), torch 2.x, macOS arm64, CPU.

---
*Filed with [Claude Code](https://claude.com/claude-code); measurements were run locally and are reproducible from the snippets above.*

**Comment by gdalsanto** [comment]
Solved with b8666c12a19206665cfb19134f22e343fc61753a
